common/protos/linux_riscv64.c

```
@@ -0,0 +1,119 @@
+        panic(true, "linux: protocols < 0.2 are not supported");
+    }
+
+    size_t kernel_size = kernel_file->size;
+    void *kernel_base = ext_mem_alloc_type_aligned(
+                ALIGN_UP(kernel_size, 4096),
+                MEMMAP_KERNEL_AND_MODULES, 2 * 1024 * 1024);
+    fread(kernel_file, kernel_base, 0, kernel_size);
+    fclose(kernel_file);
+    printv("linux: loaded kernel `%s` at %x, size %u\n", kernel_path, kernel_base, kernel_size);
+
+    void *dtb = get_device_tree_blob();
+    if (!dtb)
+        panic(true, "linux: no device tree blob found");
+
+    int ret = fdt_set_chosen_string(dtb, "bootargs", cmdline);
+    if (ret < 0)
+       printv("linux: cannot set bootargs: `%s`\n", fdt_strerror(ret));
+
+    char *module_path = config_get_value(config, 0, "MODULE_PATH");
+    if (module_path) {
+        struct file_handle *module_file = uri_open(module_path);
+        if (!module_file)
+            panic(true, "linux: failed to open module `%s`. Is the path correct?", module_path);
+
+        size_t module_size = module_file->size;
+        void *module_base = ext_mem_alloc_type_aligned(
+                        ALIGN_UP(module_size, 4096),
+                        MEMMAP_KERNEL_AND_MODULES, 4096);
+
+        fread(module_file, module_base, 0, module_size);
+        fclose(module_file);
+        printv("linux: loaded module `%s` at %x, size %u\n", module_path, module_base, module_size);
+
+        ret = fdt_set_chosen_uint64(dtb, "linux,initrd-start", (uint64_t)module_base);
+        if (ret < 0)
+            printv("cannot set initrd: %s\n", fdt_strerror(ret));
+
+        ret = fdt_set_chosen_uint64(dtb, "linux,initrd-end", (uint64_t)(module_base + module_size));
+        if (ret < 0)
+            printv("linux: cannot set initrd parameter: %s\n", fdt_strerror(ret));
+
+    }
+
+    printv("linux: bsp hart %d, device tree blob at %x\n", bsp_hartid, dtb);
+
+    void (*kernel_entry)(uint64_t hartid, uint64_t dtb) = kernel_base;
```

Should be marked as `noreturn`.

---

A function pointer cannot be marked as noreturn, only definition and declaration could. A `__builtin_unreachable()` is inserted to prevent warnings.

---

That's odd, I've been able to use at least `__attribute__((noreturn))` on pointers before, unless it was just silently ignored.

---

I am probably just misremebering, sorry.